phoenix-core/src/main/java/org/apache/phoenix/jdbc/bootstrapz/HRpcHBaseRegistryBootstrap.java

```diff
@@ -0,0 +1,44 @@
+package org.apache.phoenix.jdbc.bootstrapz;
```

Comment:
I think that moving parameter parsing to the bootstrap module, and having an API like:

```
BootstrapRegistry bs = BootstrapFactory.fromURL(String URL);
Map connpProps = bs.generateConnectionProps();

bs.toString()
bs.getZookeperQuorum() (which fails for HRpc)
bs.equals(), bs.hashMap()
```

would be a better encapsulation, and simplify code.

WDYT @joshelser (as this was originally your suggestion) ?

Reply:
It gets a little tricky because the URL is mutated accordingly to tokenize later if there's a +; and we'd be moving that logic into the `fromURL` method. Also we'd have to move the `MalformedUrlException` class into its own package to avoid a circular dependency.

Reply:
Moving MalformedUrlException to a seprate file is not a problem.

Also the I think that the URL without the boostrap info could be returned as

```
BootstrapRegistry bs = BootstrapFactory.fromURL(String URL);
Map connpProps = bs.generateConnectionProps();
url = bs.getRemainingURL();
```

Reply:
@stoty check out the updates. I don't think it makes sense to persist the URL in the bootstrap. There's no reason for the BootstrapRegistry object to keep the context of the URL... so I ended up creating a factory method that parses the URL, and spits out a BootstrapRegistry along with the remaining URL in a tuple format.

Let me know if that looks good I'll get started on the integration test with the mini cluster.

Reply:
I don't want to hold up the patch, but the current solution feels like the worst of both worlds, as it adds a lots of complexity without removing any.

Classic OO principles say that you should encapsulate your code and data in a class.
The current stateless solution is not worth it. It relies on the fact the state for the ZK and HRpp connections are named similarly, and can be stored in the same variables, but is a bad abstraction.

IMO either just fold back the code into the main class, and save the hassle of using registries, or use stateful registry classes with proper data encapsulation as I suggested above.

@joshelser @yanxinyi  what's your take on this ?

Reply:
> The current stateless solution is not worth it.

> you should encapsulate your code and data in a class.

I guess I'm failing to see the opportunity here for encapsulating the code/data. My original intent was to abstract the interface for `generateConnectionProps` and the logic each registry has for it. Doesn't necessarily mean we have to feed it some data to abstract as well. Sure maybe down the road, it may make sense to move more data from `PhoenixEmbeddedDriver` into these registry classes as per need... but I don't think any of these classes require context from the URL after being created for the current use case. 

> relies on the fact the state for the ZK and HRpp connections are named similarly

How? Was hoping to rely on the fact that unique ZK and HRpc identifiers can be parsed into a discrete set of enums.